tests/internals/cache_metrics/cache_metrics_test.go

```diff
@@ -0,0 +1,224 @@
+func TestScaler(t *testing.T) {
+	// setup
+	t.Log("--- setting up ---")
+	// Create kubernetes resources
+	kc := GetKubernetesClient(t)
+	data, templates := getTemplateData()
+
+	CreateKubernetesResources(t, kc, testNamespace, data, templates)
+
+	// test direct metrics query (the standard behavior)
+	testDirectQuery(t, kc, data)
+
+	// test querying metrics on polling interval
+	testCacheMetricsOnPollingInterval(t, kc, data)
+
+	// cleanup
+	DeleteKubernetesResources(t, kc, testNamespace, data, templates)
+}
+
+func getTemplateData() (templateData, []Template) {
+	return templateData{
+			TestNamespace:               testNamespace,
+			DeploymentName:              deploymentName,
+			ScaledObjectName:            scaledObjectName,
+			MonitoredDeploymentName:     monitoredDeploymentName,
+			MonitoredDeploymentReplicas: defaultMonitoredDeploymentReplicas,
+		}, []Template{
+			{Name: "deploymentTemplate", Config: deploymentTemplate},
+			{Name: "monitoredDeploymentTemplate", Config: monitoredDeploymentTemplate},
+		}
+}
+
+func testCacheMetricsOnPollingInterval(t *testing.T, kc *kubernetes.Clientset, data templateData) {
+	t.Log("--- testing caching metrics on polling interval ---")
+
+	KubectlApplyWithTemplate(t, data, "deploymentTemplate", deploymentTemplate)
+	data.MonitoredDeploymentReplicas = defaultMonitoredDeploymentReplicas
+	KubectlApplyWithTemplate(t, data, "monitoredDeploymentTemplate", monitoredDeploymentTemplate)
+
+	// initial replica count for a deployment
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 1, 60, 1),
+		"replica count should be 1 after 1 minute")
+
+	// initial replica count for a monitored deployment
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, monitoredDeploymentName, testNamespace, defaultMonitoredDeploymentReplicas, 60, 1),
+		fmt.Sprintf("replica count should be %d after 1 minute", defaultMonitoredDeploymentReplicas))
+
+	data.EnableUseCachedMetrics = true
+	KubectlApplyWithTemplate(t, data, "scaledObjectTemplate", scaledObjectTemplate)
+
+	// Metric Value = 8, DesiredAverageMetricValue = 2
+	// should scale in to 8/2 = 4 replicas, irrespective of current replicas
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, 4, 60, 1),
+		"replica count should be 4 after 1 minute")
+
+	// Changing Metric Value to 4, but because we have a long polling interval, the replicas number should remain the same
+	data.MonitoredDeploymentReplicas = 4
+	KubectlApplyWithTemplate(t, data, "monitoredDeploymentTemplate", monitoredDeploymentTemplate)
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, monitoredDeploymentName, testNamespace, 4, 60, 1),
+		fmt.Sprintf("replica count should be %d after 1 minute", defaultMonitoredDeploymentReplicas))
+
+	t.Log("waiting 30s to allow propagation of changes")
+
+	// Let's sleep at least 30s
+	// the usual setting for `horizontal-pod-autoscaler-sync-period` is 15s)
+	time.Sleep(30 * time.Second)
```

Comment:
I think we can use the helper function `AssertReplicaCountNotChangeDuringTimePeriod` to ensure that replica count doesn't change. Internally, the helper waits some time and evaluates the deployment replica count. WDYT?